File: converted_source/Uundo.pas.py
# unit Uundo

from conversion_common import *
import umain
import udomain
import umath
import updform
import ucommand
import delphi_compatability

# const
kBetweenGap = 4

class TUndoRedoListForm(PdForm):
    def __init__(self):
        self.OK = TButton()
        self.cancel = TButton()
        self.helpButton = TSpeedButton()
        self.undoListLabel = TLabel()
        self.redoListLabel = TLabel()
        self.undoList = TListBox()
        self.redoList = TListBox()
    
    #$R *.DFM
    def FormCreate(self, Sender):
        tempBoundsRect = TRect()
        
        # keep window on screen - left corner of title bar 
        tempBoundsRect = udomain.domain.undoRedoListWindowRect
        if (tempBoundsRect.Left != 0) or (tempBoundsRect.Right != 0) or (tempBoundsRect.Top != 0) or (tempBoundsRect.Bottom != 0):
            if tempBoundsRect.Left > delphi_compatability.Screen.Width - umain.kMinWidthOnScreen:
                tempBoundsRect.Left = delphi_compatability.Screen.Width - umain.kMinWidthOnScreen
            if tempBoundsRect.Top > delphi_compatability.Screen.Height - umain.kMinHeightOnScreen:
                tempBoundsRect.Top = delphi_compatability.Screen.Height - umain.kMinHeightOnScreen
            self.SetBounds(tempBoundsRect.Left, tempBoundsRect.Top, tempBoundsRect.Right, tempBoundsRect.Bottom)
        # make list boxes have horizontal scroll bars
        UNRESOLVED.sendMessage(self.undoList.Handle, UNRESOLVED.LB_SETHORIZONTALEXTENT, 1000, 0)
        UNRESOLVED.sendMessage(self.redoList.Handle, UNRESOLVED.LB_SETHORIZONTALEXTENT, 1000, 0)
    
    def initializeWithCommandList(self, commandList):
        aList = TStringList()
        
        aList = delphi_compatability.TStringList.create
        try:
            commandList.fillListWithUndoableStrings(aList)
            self.undoList.Items.Assign(aList)
            aList.Clear()
            commandList.fillListWithRedoableStrings(aList)
            self.redoList.Items.Assign(aList)
        finally:
            aList.free
        self.OK.Enabled = (self.undoList.SelCount > 0) or (self.redoList.SelCount > 0)
    
    def undoToIndex(self):
        result = 0
        i = 0
        
        result = -1
        if self.undoList.Items.Count > 0:
            for i in range(self.undoList.Items.Count - 1, 0 + 1):
                if self.undoList.Selected[i]:
                    result = i
                    return result
        return result
    
    def redoToIndex(self):
        result = 0
        i = 0
        
        result = -1
        if self.redoList.Items.Count > 0:
            for i in range(self.redoList.Items.Count - 1, 0 + 1):
                if self.redoList.Selected[i]:
                    result = i
                    return result
        return result
    
    def regulateListSelections(self, Sender):
        i = 0
        thisListBox = TListBox()
        otherListBox = TListBox()
        
        if (Sender == None) or (not (Sender.__class__ is delphi_compatability.TListBox)):
            raise GeneralException.create("Problem: Wrong sender in method TUndoRedoListForm.regulateListSelections.")
        thisListBox = Sender as delphi_compatability.TListBox
        if Sender == self.undoList:
            otherListBox = self.redoList
        else:
            otherListBox = self.undoList
        if Sender == self.undoList:
            self.OK.Caption = "Un&do"
        else:
            self.OK.Caption = "Re&do"
        self.OK.Enabled = (self.undoList.SelCount > 0) or (self.redoList.SelCount > 0)
        if thisListBox.ItemIndex >= 0:
            for i in range(0, thisListBox.ItemIndex + 1):
                thisListBox.Selected[i] = true
        if thisListBox.ItemIndex + 1 <= thisListBox.Items.Count - 1:
            for i in range(thisListBox.ItemIndex + 1, thisListBox.Items.Count):
                thisListBox.Selected[i] = false
        if otherListBox.Items.Count > 0:
            for i in range(0, otherListBox.Items.Count):
                otherListBox.Selected[i] = false
        otherListBox.ItemIndex = -1
    
    def OKClick(self, Sender):
        self.ModalResult = mrOK
    
    def cancelClick(self, Sender):
        self.ModalResult = mrCancel
    
    def FormClose(self, Sender, Action):
        udomain.domain.undoRedoListWindowRect = Rect(self.Left, self.Top, self.Width, self.Height)
    
    def helpButtonClick(self, Sender):
        delphi_compatability.Application.HelpJump("Using_the_undo/redo_list")
    
    def FormResize(self, Sender):
        listHeight = 0
        
        if delphi_compatability.Application.terminated:
            return
        self.OK.Left = self.ClientWidth - self.OK.Width - kBetweenGap
        self.cancel.Left = self.OK.Left
        self.helpButton.Left = self.OK.Left
        self.undoListLabel.SetBounds(kBetweenGap, kBetweenGap, self.undoListLabel.Width, self.undoListLabel.Height)
        listHeight = umath.intMax(0, (self.ClientHeight - self.undoListLabel.Height * 2 - kBetweenGap * 5) / 2)
        self.undoList.SetBounds(kBetweenGap, self.undoListLabel.Top + self.undoListLabel.Height + kBetweenGap, self.OK.Left - kBetweenGap * 2, listHeight)
        self.redoListLabel.SetBounds(kBetweenGap, self.undoList.Top + self.undoList.Height + kBetweenGap, self.redoListLabel.Width, self.redoListLabel.Height)
        self.redoList.SetBounds(kBetweenGap, self.redoListLabel.Top + self.redoListLabel.Height + kBetweenGap, self.OK.Left - kBetweenGap * 2, listHeight)
    
    def WMGetMinMaxInfo(self, MSG):
        PdForm.WMGetMinMaxInfo(self)
        # FIX unresolved WITH expression: UNRESOLVED.PMinMaxInfo(MSG.lparam).PDF_FIX_POINTER_ACCESS
        UNRESOLVED.ptMinTrackSize.x = 220
        UNRESOLVED.ptMinTrackSize.y = 220
    

File: converted_source/Uundo.pas.java
// unit Uundo

from conversion_common import *;
import updform;
import delphi_compatability;

// const
kBetweenGap = 4;



class TUndoRedoListForm extends PdForm {
    public TButton OK;
    public TButton cancel;
    public TSpeedButton helpButton;
    public TLabel undoListLabel;
    public TLabel redoListLabel;
    public TListBox undoList;
    public TListBox redoList;
    
    //$R *.DFM
    public void FormCreate(TObject Sender) {
        TRect tempBoundsRect = new TRect();
        
        // keep window on screen - left corner of title bar 
        tempBoundsRect = UNRESOLVED.domain.undoRedoListWindowRect;
        if ((tempBoundsRect.Left != 0) || (tempBoundsRect.Right != 0) || (tempBoundsRect.Top != 0) || (tempBoundsRect.Bottom != 0)) {
            if (tempBoundsRect.Left > delphi_compatability.Screen.Width - UNRESOLVED.kMinWidthOnScreen) {
                tempBoundsRect.Left = delphi_compatability.Screen.Width - UNRESOLVED.kMinWidthOnScreen;
            }
            if (tempBoundsRect.Top > delphi_compatability.Screen.Height - UNRESOLVED.kMinHeightOnScreen) {
                tempBoundsRect.Top = delphi_compatability.Screen.Height - UNRESOLVED.kMinHeightOnScreen;
            }
            this.SetBounds(tempBoundsRect.Left, tempBoundsRect.Top, tempBoundsRect.Right, tempBoundsRect.Bottom);
        }
        // make list boxes have horizontal scroll bars
        UNRESOLVED.sendMessage(this.undoList.Handle, UNRESOLVED.LB_SETHORIZONTALEXTENT, 1000, 0);
        UNRESOLVED.sendMessage(this.redoList.Handle, UNRESOLVED.LB_SETHORIZONTALEXTENT, 1000, 0);
    }
    
    public void initializeWithCommandList(PdCommandList commandList) {
        TStringList aList = new TStringList();
        
        aList = delphi_compatability.TStringList.create;
        try {
            commandList.fillListWithUndoableStrings(aList);
            this.undoList.Items.Assign(aList);
            aList.Clear();
            commandList.fillListWithRedoableStrings(aList);
            this.redoList.Items.Assign(aList);
        } finally {
            aList.free;
        }
        this.OK.Enabled = (this.undoList.SelCount > 0) || (this.redoList.SelCount > 0);
    }
    
    public int undoToIndex() {
        result = 0;
        int i = 0;
        
        result = -1;
        if (this.undoList.Items.Count > 0) {
            for (i = this.undoList.Items.Count - 1; i >= 0; i--) {
                if (this.undoList.Selected[i]) {
                    result = i;
                    return result;
                }
            }
        }
        return result;
    }
    
    public int redoToIndex() {
        result = 0;
        int i = 0;
        
        result = -1;
        if (this.redoList.Items.Count > 0) {
            for (i = this.redoList.Items.Count - 1; i >= 0; i--) {
                if (this.redoList.Selected[i]) {
                    result = i;
                    return result;
                }
            }
        }
        return result;
    }
    
    public void regulateListSelections(TObject Sender) {
        int i = 0;
        TListBox thisListBox = new TListBox();
        TListBox otherListBox = new TListBox();
        
        if ((Sender == null) || (!(Sender instanceof delphi_compatability.TListBox))) {
            throw new GeneralException.create("Problem: Wrong sender in method TUndoRedoListForm.regulateListSelections.");
        }
        thisListBox = (delphi_compatability.TListBox)Sender;
        if (Sender == this.undoList) {
            otherListBox = this.redoList;
        } else {
            otherListBox = this.undoList;
        }
        if (Sender == this.undoList) {
            this.OK.Caption = "Un&do";
        } else {
            this.OK.Caption = "Re&do";
        }
        this.OK.Enabled = (this.undoList.SelCount > 0) || (this.redoList.SelCount > 0);
        if (thisListBox.ItemIndex >= 0) {
            for (i = 0; i <= thisListBox.ItemIndex; i++) {
                thisListBox.Selected[i] = true;
            }
        }
        if (thisListBox.ItemIndex + 1 <= thisListBox.Items.Count - 1) {
            for (i = thisListBox.ItemIndex + 1; i <= thisListBox.Items.Count - 1; i++) {
                thisListBox.Selected[i] = false;
            }
        }
        if (otherListBox.Items.Count > 0) {
            for (i = 0; i <= otherListBox.Items.Count - 1; i++) {
                otherListBox.Selected[i] = false;
            }
        }
        otherListBox.ItemIndex = -1;
    }
    
    public void OKClick(TObject Sender) {
        this.ModalResult = mrOK;
    }
    
    public void cancelClick(TObject Sender) {
        this.ModalResult = mrCancel;
    }
    
    public void FormClose(TObject Sender, TCloseAction Action) {
        UNRESOLVED.domain.undoRedoListWindowRect = Rect(this.Left, this.Top, this.Width, this.Height);
    }
    
    public void helpButtonClick(TObject Sender) {
        delphi_compatability.Application.HelpJump("Using_the_undo/redo_list");
    }
    
    public void FormResize(TObject Sender) {
        int listHeight = 0;
        
        if (delphi_compatability.Application.terminated) {
            return;
        }
        this.OK.Left = this.ClientWidth - this.OK.Width - kBetweenGap;
        this.cancel.Left = this.OK.Left;
        this.helpButton.Left = this.OK.Left;
        this.undoListLabel.SetBounds(kBetweenGap, kBetweenGap, this.undoListLabel.Width, this.undoListLabel.Height);
        listHeight = UNRESOLVED.intMax(0, (this.ClientHeight - this.undoListLabel.Height * 2 - kBetweenGap * 5) / 2);
        this.undoList.SetBounds(kBetweenGap, this.undoListLabel.Top + this.undoListLabel.Height + kBetweenGap, this.OK.Left - kBetweenGap * 2, listHeight);
        this.redoListLabel.SetBounds(kBetweenGap, this.undoList.Top + this.undoList.Height + kBetweenGap, this.redoListLabel.Width, this.redoListLabel.Height);
        this.redoList.SetBounds(kBetweenGap, this.redoListLabel.Top + this.redoListLabel.Height + kBetweenGap, this.OK.Left - kBetweenGap * 2, listHeight);
    }
    
    public void WMGetMinMaxInfo(Tmessage MSG) {
        super.WMGetMinMaxInfo();
        //FIX unresolved WITH expression: UNRESOLVED.PMinMaxInfo(MSG.lparam).PDF_FIX_POINTER_ACCESS
        UNRESOLVED.ptMinTrackSize.x = 220;
        UNRESOLVED.ptMinTrackSize.y = 220;
    }
    
}
